arangod/Agency/Supervision.cpp

```diff
@@ -3496,3 +3502,207 @@ void Supervision::removeTransactionBuilder(
+        auto inPending = _snapshot->hasAsNode(pendingPrefix + jobId.value());
+        if (!inTodo && !inPending) {
+          return true;
+        }
+      } else {
+        return true;
+        // This should not happen, since if started is there, we have a
+        // jobId.
+      }
+    }
+    auto jobOpt = entry->hasAsNode("moveShard");
+    if (!jobOpt) {
+      return true;
+    }
+    Node const& job(jobOpt.value());
+    auto database = job.hasAsString("database");
+    auto collection = job.hasAsString("collection");
+    auto server = job.hasAsString("fromServer");
+    if (!server) {
+      return true;
+    } else {
+      if (isReplication2(*database)) {
+        auto stateId = LogicalCollection::shardIdToStateId(shard);
+        auto target = Job::readStateTarget(snapshot(), *database, stateId);
+        if (not target->participants.contains(*server)) {
+          return true;
+        }
+      } else {
+        if (!isServerInPlan(database.value(), collection.value(), shard,
+                            server.value())) {
+          return true;
+        }
+      }
+    }
+    return false;
+  };
+
+  auto const reclaimShard =
+      [&](std::shared_ptr<VPackBuilder> const& trx, std::string const& database,
+          std::string const& collection, std::string const& shard,
+          std::string const& newLeader, std::string const& oldLeader) {
+        uint64_t jobId = _jobId++;
+        MoveShard(*_snapshot, _agent, std::to_string(jobId), "supervision",
+                  database, collection, shard, newLeader, oldLeader, true, true,
+                  false)
+            .create(trx);
+
+        std::string now(timepointToString(std::chrono::system_clock::now()));
+        std::string path = returnLeadershipPrefix + std::string(shard) + "/";
+        trx->add(path + "started", VPackValue(now));
+        trx->add(path + "jobId", VPackValue(std::to_string(jobId)));
+      };
+
+  auto undos = snapshot().hasAsChildren("Target/ReturnLeadership");
+  if (not undos) {
+    return;
+  }
+
+  // Collect a transaction:
+  auto trx = std::make_shared<VPackBuilder>();
+  {
+    VPackArrayBuilder guard(trx.get());
+    VPackObjectBuilder guard2(trx.get());
+
+    for (auto const& [shard, entry] : undos->get()) {
+      // First check some conditions under which we simply get rid of the
+      // entry:
+      //  - deadline exceeded (and not yet started)
+      //  - dependent MoveShard gone (and started)
+      //  - fromServer no longer in Plan
+      if (checkDeletion(shard, entry)) {
+        // Let's remove the entry:
+        trx->add(VPackValue(returnLeadershipPrefix + shard));
+        {
+          VPackObjectBuilder guard3(trx.get());
+          trx->add("op", VPackValue("delete"));
+        }
+        continue;
+      }
+
+      // Now the job is still valid, let's see if it fires. Note that we
+      // do not get here, if the job is not found!
+      auto jobOpt = entry->hasAsNode("moveShard");
+      TRI_ASSERT(jobOpt);
```

Comment:
Can't tell if this assertion is true for old jobs as well.

Reply:
Again, this is only for `MoveShard` jobs, which show up in the undo section `/arango/Target/ReturnLeadership/<shard>` under `moveShard`. These can only exist if they were generated from the code in this PR. Besides, the `checkDeletion` code a few lines above has already verified existence of the entry. So this should be OK, too.

Reply:
Ok!